tests/file_storage/tests.py

```diff
@@ -1157,3 +1160,288 @@ def test_urllib_request_urlopen(self):
         remote_file = urlopen(self.live_server_url + "/")
         with self.storage.open(stored_filename) as stored_file:
             self.assertEqual(stored_file.read(), remote_file.read())
+
+
+class MemoryStorageIOTestCase(unittest.TestCase):
+    def setUp(self):
+        self.storage = InMemoryStorage()
+
+    def test_write_string(self):
+        """
+        Basic I/O operations should work as expected
+        """
+        with self.storage.open("file.txt", "w") as fd:
+            fd.write("hello")
+
+        with self.storage.open("file.txt", "r") as fd:
+            self.assertEqual(fd.read(), "hello")
+
+        with self.storage.open("file.dat", "wb") as fd:
+            fd.write(b"hello")
+
+        with self.storage.open("file.dat", "rb") as fd:
+            self.assertEqual(fd.read(), b"hello")
+
+    def test_convert_str_to_bytes_and_back(self):
+        """
+        InMemoryStorage should handle conversion from str to bytes and back
+        """
+        with self.storage.open("file.txt", "w") as fd:
+            fd.write("hello")
+
+        with self.storage.open("file.txt", "rb") as fd:
+            self.assertEqual(fd.read(), b"hello")
+
+        with self.storage.open("file.dat", "wb") as fd:
+            fd.write(b"hello")
+
+        with self.storage.open("file.dat", "r") as fd:
+            self.assertEqual(fd.read(), "hello")
+
+    def test_open_missing_file(self):
+        """
+        Opening a missing file should raise a FileNotFoundError
+        """
+        self.assertRaises(FileNotFoundError, self.storage.open, "missing.txt")
+
+    def test_open_dir_as_file(self):
+        """
+        Opening a dir as file should raise an IsADirectoryError
+        """
+        with self.storage.open("a/b/file.txt", "w") as fd:
+            fd.write("hello")
+        self.assertRaises(IsADirectoryError, self.storage.open, "a/b")
+
+    def test_file_saving(self):
+        """
+        Basic file saving should work as expected
+        """
+        self.storage.save("file.txt", ContentFile("test"))
+        self.assertEqual(self.storage.open("file.txt", "r").read(), "test")
+
+        self.storage.save("file.dat", ContentFile(b"test"))
+        self.assertEqual(self.storage.open("file.dat", "rb").read(), b"test")
+
+    @unittest.skipIf(
+        sys.platform == "win32", "Windows doesn't support moving open files."
+    )
+    def test_removing_temporary_file_after_save(self):
+        """
+        A temporary file should be removed when saved into storage
+        """
+        with TemporaryUploadedFile("test", "text/plain", 1, "utf8") as file:
+            self.storage.save("test.txt", file)
+            self.assertFalse(os.path.exists(file.temporary_file_path()))
```

Comment:
As far as I'm aware this is already covered by `test_move_temporary_file`. I think we can remove it :thinking: